### DISCUSSION TITLE
Implementing Custom Kernels (Mixed Formulation) with Heat Conduction as Example

### DISCUSSION THREAD

**Discussion by richmondodufisan** [idaholab/moose#26021] [Q&A General] [answered]
Hi, I wanted to know the proper syntax for implementing this. I'm not getting converging results for this simple problem, and so I wanted to be sure my syntax was/is correct. Starting with the equations and their corresponding finite element discretizations:

**Diffusion equation with variation taken wrt to T**:

![diffusion](https://github.com/idaholab/moose/assets/52802063/fe225d65-c7c9-4f0c-a7fa-139920b7110e)

Considering the steady state case first, in my Kernel Header, I have this declaration:

![1](https://github.com/idaholab/moose/assets/52802063/c8213c5d-880b-45dc-bd9a-eb209aebbf94)

And the residual computed like this:

![2](https://github.com/idaholab/moose/assets/52802063/2b0da9b5-b509-4016-ab89-43ce340a1172)

**A Fourier relationship between Flux and Temperature with variation taken wrt to q**:

![fourier](https://github.com/idaholab/moose/assets/52802063/0f11ab6a-045d-4ce8-92fc-ad16268e7437)

In the Kernel Header, I have:

![3](https://github.com/idaholab/moose/assets/52802063/1d58b9b4-9e3c-45e6-8d2d-2d5c91e47a8e)

And the residual computation:

![4](https://github.com/idaholab/moose/assets/52802063/094c4dc4-6767-4283-85d9-83ffce2f8ccb)

**Kernels block in input file**

![5](https://github.com/idaholab/moose/assets/52802063/fe5d06dc-2296-4fb9-9613-4529f1a70bdd)

To my understanding this has been implemented correctly, but its not converging for what should be a very simple problem (a box with temperature ranging from 0 to 1) and so clearly I'm doing something wrong. Simulations failed with first order elements for all variables, and also with second order q variables. Could you please take a look?

**Comment by lindsayad** [comment]
Hi @richmondodufisan glancing over this, this looks good to me. If you wanted to share a github branch with your source and input, I could test it out and see if I can identify something problematic

**Reply by richmondodufisan** [reply]
@lindsayad thanks! Here's a link to the repository: https://github.com/richmondodufisan/FDTR_Simulation_Hydrodynamics

The input file is in Formulation_Test/Simple_Box

There is a script to replace the app name in all the source files too to make that easier.

**Reply by lindsayad** [reply]
How do I `make` your application?

**Reply by richmondodufisan** [reply]
Oh the whole application isn't what's on github, maybe that's what I should be doing instead. I just thought it would be redundant to create a new app anytime I wanted to add new but different physics. So instead I just created a repo with the custom objects for any user to be able to download and add to their app.

The repository I sent you has the source and header files (in the Kernels_and_Headers folder) needed which are to be copied to yourAppName/src/kernels and yourAppName/include/kernels respectively. And then you can `make`.

 I added a script to change the app name before copying with usage `Usage: ./Replace_App.sh <old_app_name> <new_app_name>` where in this case old_app_name (name of my app) is "purple"

**Reply by lindsayad** [reply]
It would be better to create your own app. It can always be easily incorporated into others applications. For instance https://github.com/arfc/squirrel is an "app" but it's pulled in and used as a library by multiple other apps such as zapdos and moltres

**Reply by lindsayad** [reply, marked as answer]
So if you use `-pc_type lu` you will get an answer but it will look funny as shown below. What you need to add now are "implicit" boundary conditions to model the boundary terms coming from integration by parts. Right now you are implicitly setting the boundary terms for the "q equation" to 0 at all boundaries and for the "T equation" to 0 at the top and bottom boundaries

<img width="770" alt="Screenshot 2023-11-13 at 1 21 01 PM" src="https://github.com/idaholab/moose/assets/10248304/d5029242-3382-420e-b848-1ecb1637ed91">

**Reply by lindsayad** [reply]
You will have no hope of getting a solution with boomeramg BTW for the Steady case since there is no diagonal for the temperature equation

**Reply by richmondodufisan** [reply]
Hmm I see, Will it make sense then, to not integrate the temperature term by parts in the q equation? Since the only flux/boundary term I actually want to use is the one from the t equation (i.e q.n = 0 on all natural boundaries)? I did try that but it still didn't converge. Or what boundary conditions would be closest to the usual formulation in ADHeatConduction?

Also, to be honest I don't know what most of those options in the Executioner mean, I should probably look into them further. I copied them from another input file when I was initially learning to use MOOSE with the basic understanding that if I used `solve_type = 'PJFNK'` and made my custom physics inherit from ADKernel, I wouldn't need to code my Jacobians myself thanks to automatic differentiation. I haven't looked in detail at what the rest of the options mean apart from the tolerance. 

created the repository btw: https://github.com/richmondodufisan/purple

**Reply by lindsayad** [reply]
> Hmm I see, Will it make sense then, to not integrate the temperature term by parts in the q equation?

Yes you can do that.

> I did try that but it still didn't converge

Did you try it with `-pc_type lu`?

**Reply by richmondodufisan** [reply]
Actually, I tried it again (not integrating the T term in the q equation by parts) and this time it worked! The reason you gave made so much sense and I'm glad it converged, thank you! So it turns out it was something about those other options that was the reason it didn't converge before.

![heat_eqn](https://github.com/idaholab/moose/assets/52802063/01231010-cdf9-4ae5-9b70-d198037f994b)

One thing that I did notice though, is the "outlier variable residual norms" in red. What does that mean?

```
 0 Nonlinear |R| = 3.060206e+00
 1 Nonlinear |R| = 3.005904e-05
 2 Nonlinear |R| = 2.818184e-10
 Solve Converged!
Currently Executing
  Finished Solving                                                                       [  6.71 s] [  157 MB]

Outlier Variable Residual Norms:
  temperature: 2.803203e-10
```

**Reply by lindsayad** [reply]
> So it turns out it was something about those other options that was the reason it didn't converge before.

Yea boomeramg works best for symmetric positive definite matrices and your system is indefinite due to the missing diagonal in the T equation so boomeramg will not work well. Generally when debugging I always use `-pc_type lu` since that is basically a direct inversion of the preconditioning matrix, so it is the strongest preconditioner you can apply and will work with all systems. It just has very poor efficiency for large matrices.

> One thing that I did notice though, is the "outlier variable residual norms" in red. What does that mean?

It just means that it's residual is larger than the other variable by a significant margin. But you've dropped the overall residual norm by 10 orders of magnitude so I would not worry about it.

**Reply by richmondodufisan** [reply]
I see I see that makes sense, thank you! Learned a lot here, really appreciate your help.

**Reply by lindsayad** [reply]
You're welcome!

**Reply by richmondodufisan** [reply]
@lindsayad another question- what is the convention for the normal vector in MOOSE? Is it pointing outward or inward? Asking because I just tried two sets of simulations where in order to get the same results, I had to do the following:

-In the first one, I integrated the q term in the "T equation" to have a boundary term in T. Then I specified a (positive) Neumann condition of 1000 at the right boundary. 

-In the second one, I did not integrate the T term, and instead applied a Dirichlet condition on the right side x-component of q of -1000. Setting it to +1000 gave me the same "spread" but negative temperatures which is obviously unphysical.

The right direction points in the direction of the x-axis, and so it made sense why a negative dirichlet condition would be necessary there (i.e to make the flux point in the direction of the mesh). But then it also made me wonder how MOOSE has been handling the normal vector.

As a side note, I was doing this because my larger problem was not converging but I have narrowed down the issue to the order of the elements I was using- I tried to use first instead of second order elements for q because my meshing software was producing inverted elements (= negative jacobian) at the interface with the second order tag on. But using first order elements made the simulations unable to converge past the first two timesteps. For the simple box case, first order elements did not converge at all.

**Reply by lindsayad** [reply]
You will definitely need higher order basis for q than for T. This is due to the missing diagonal in the T equation.

The normal vector always points outwards. But if you look at `NeumannBC` there is no use of the `_normals` so you have to be careful with the sign of `value` that you provide.

> my meshing software was producing inverted elements

You could use your meshing software to create first order elements and then if you specify `second_order = true` in your mesh block, then MOOSE/libMesh will turn your first order elements into second order for you

**Reply by richmondodufisan** [reply]
I see I see. MOOSE really can do it all, wow.

**Reply by lindsayad** [reply]
Haha I wouldn't say that, but I'm happy to say we can do a fair amount!
